### PR TITLE
Allow for IDs to be strings as well as numbers and also not require them to sequential

### PR DESCRIPTION
**Pull request by kpustakhod**
What was doen in this PR:
1. Refactored the way to access particular node in data. Originally data is an `INode[]` with sequential numbers as data indexes and node ids and node access was implemented as `data[id]`. Now data is still `INode[]` but supports non-sequential ids and node access is implemented with `getTreeParent` and `getTreeNode` helpers.
2. Node ids now supports numbers and strings.
3. Added additional validation for `data: INode[]` to now allow rendering of broken data. For example: two nodes are market as root nodes (`parent: null`), or there are duplicates in node's children, etc.
4. Types and reducer was moved in separate files to improve readability.

Reference: UIEN-3971, UIEN-3972

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,63 +16,67 @@ A react component that implements the treeview pattern as described by the [WAI-
...
+| `expandedIds`            | `array`                       | `[]`          | (Controlled) Array with the ids of branch node that should be expanded                                                                                                    |
+| `clickAction`            | `enum`                        | `SELECT`      | Action to perform on click. One of: EXCLUSIVE_SELECT, FOCUS, SELECT                                                                                                       |
+| `onBlur`                 | `func`                        | `noop`        | Custom onBlur event that is triggered when focusing out of the component as a whole (moving focus between the nodes won't trigger it).                                    |
 
 <br/> <br/>
 
 ## data
 
-An array of nodes. Nodes are objects with the following structure:
+Data can be eigther array of nodes or Map of nodes with unique and non-sequential ids.
```
Comment (MehmetYararVX) [inline README.md]: typo: eigther

File: src/TreeView/README.md [hunk 2]
```diff
@@ -10,62 +10,66 @@ A react component that implements the treeview pattern as described by the [WAI-
...
+| `selectedIds`            | `array`                       | `[]`          | (Controlled) Array with the ids that should be selected                                                                                                                   |
+| `expandedIds`            | `array`                       | `[]`          | (Controlled) Array with the ids of branch node that should be expanded                                                                                                    |
+| `clickAction`            | `enum`                        | `SELECT`      | Action to perform on click. One of: EXCLUSIVE_SELECT, FOCUS, SELECT                                                                                                       |
 
 <br/> <br/>
 
 ### data
 
-An array of nodes. Nodes are objects with the following structure:
+Data can be eigther array of nodes or Map of nodes with unique and non-sequential ids.
```
Comment (MehmetYararVX) [inline src/TreeView/README.md]: typo: eigther

File: src/TreeView/utils.ts [hunk 3]
```diff
@@ -26,28 +26,30 @@ export const difference = (a: Set<number>, b: Set<number>) => {
...
     ref.current = value;
   });
   return ref.current;
 };
 
-export const isBranchNode = (data: INode[], i: number) =>
-  data[i].children != null && data[i].children.length > 0;
+export const isBranchNode = (data: TreeViewData, i: NodeId) => {
+  const node = getTreeNode(data, i);
+  return node.children != null && node.children.length > 0;
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts]: would optional chaining work? `!!node.children?.length`

File: src/TreeView/utils.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import { useEffect, useRef } from "react";
-import { INode, INodeRef } from ".";
+import { INode, INodeRef, NodeId, TreeViewData } from ".";
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts]: I'd put all the types in a separate type file (e.g., `types.ts`).

File: src/TreeView/utils.ts [hunk 5]
```diff
@@ -26,28 +26,30 @@ export const difference = (a: Set<number>, b: Set<number>) => {
...
-export const symmetricDifference = (a: Set<number>, b: Set<number>) => {
-  return new Set<number>([...difference(a, b), ...difference(b, a)]);
+export const symmetricDifference = (a: Set<NodeId>, b: Set<NodeId>) => {
+  return new Set<NodeId>([...difference(a, b), ...difference(b, a)]);
 };
 
-export const usePrevious = (x: Set<number>): Set<number> | undefined => {
-  const ref = useRef<Set<number> | undefined>();
+export const usePrevious = (x: Set<NodeId>): Set<NodeId> | undefined => {
+  const ref = useRef<Set<NodeId> | undefined>();
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts]: I think there's no need for the `undefined` type.

File: src/TreeView/utils.ts [hunk 6]
```diff
@@ -26,28 +26,30 @@ export const difference = (a: Set<number>, b: Set<number>) => {
...
   useEffect(() => {
     ref.current = x;
   }, [x]);
   return ref.current;
 };
 
-export const usePreviousData = (value: INode[] | undefined) => {
-  const ref = useRef<INode[] | undefined>();
+export const usePreviousData = (value: TreeViewData | undefined) => {
+  const ref = useRef<TreeViewData | undefined>();
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts]: I think there's no need for the `undefined` type.

File: src/TreeView/utils.ts [hunk 7]
```diff
@@ -120,27 +122,30 @@ export const getSibling = (data: INode[], id: number, diff: number) => {
...
 export const getPreviousAccessible = (
-  data: INode[],
-  id: number,
-  expandedIds: Set<number>
+  data: TreeViewData,
+  id: NodeId,
+  expandedIds: Set<NodeId>
 ) => {
-  if (id === data[0].children[0]) {
+  if (id === getTreeParent(data).children[0]) {
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts:144]: what if there's no element in the children?

File: src/TreeView/utils.ts [hunk 8]
```diff
@@ -151,13 +156,13 @@ export const getPreviousAccessible = (
...
-  expandedIds: Set<number>
+  data: TreeViewData,
+  id: NodeId,
+  expandedIds: Set<NodeId>
 ) => {
-  let nodeId: number | null = data[id].id;
+  let nodeId: NodeId | null = getTreeNode(data, id).id;
   if (isBranchNode(data, nodeId) && expandedIds.has(nodeId)) {
-    return data[nodeId].children[0];
+    return getTreeNode(data, nodeId).children[0];
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts:161]: what if there's no element in the children?

File: src/TreeView/utils.ts [hunk 9]
```diff
@@ -275,30 +280,41 @@ export const getAccessibleRange = ({
...
-    flattenedTree[count] = node;
-    count += 1;
+
+    if (flattenedTree.has(node.id)) {
+      throw Error("TreeView node must has unique ids");
+    }
+
+    flattenedTree.set(node.id, node);
+    internalCount += 1;
     if (tree.children == null || tree.children.length === 0) return;
```
Comment (MehmetYararVX) [inline src/TreeView/utils.ts]: `!tree.children?.length`

File: src/__tests__/MultiSelectCheckboxTreeAsync.test.tsx [hunk 10]
```diff
@@ -46,20 +48,30 @@ describe("Async MultiSelectCheckboxTree", () => {
...
+    const updateTreeData = (
+      list: TreeViewData,
+      id: NodeId,
+      children: INode[]
+    ) => {
+      const data = Array.from(list).map(([,node]) => {
         if (node.id === id) {
           node.children = children.map((el) => {
             return el.id;
           });
```
Comment (MehmetYararVX) [inline src/__tests__/MultiSelectCheckboxTreeAsync.test.tsx]: `node.children = children.map(el => el.id);`

File: src/TreeView/utils.ts (hunk 8 above)
Comment (kpustakhod) [reply src/TreeView/utils.ts:161]: There is existing checks in index.tsx for this case (null/undefined).

File: src/TreeView/utils.ts (hunk 7 above)
Comment (kpustakhod) [reply src/TreeView/utils.ts:144]: Children are always array. Reference to `children[0]` won't cause any error, it will return `undefined`.

**Review by mellis481 — APPROVED** [review]

**Pull request merged**